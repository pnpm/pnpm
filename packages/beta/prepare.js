const fs = require('fs')
const path = require('path')

fs.writeFileSync(path.join(__dirname, 'pnpm'), '', 'utf8')
