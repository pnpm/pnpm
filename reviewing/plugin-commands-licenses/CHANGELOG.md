# @pnpm/plugin-commands-licenses

## 3.0.7

### Patch Changes

- @pnpm/cli-utils@2.1.4

## 3.0.6

### Patch Changes

- @pnpm/cli-utils@2.1.3

## 3.0.5

### Patch Changes

- @pnpm/license-scanner@2.2.5

## 3.0.4

### Patch Changes

- @pnpm/config@20.1.2
- @pnpm/license-scanner@2.2.4
- @pnpm/cli-utils@2.1.2

## 3.0.3

### Patch Changes

- @pnpm/license-scanner@2.2.3

## 3.0.2

### Patch Changes

- Updated dependencies [7d65d901a]
  - @pnpm/store-path@8.0.1
  - @pnpm/license-scanner@2.2.2
  - @pnpm/config@20.1.1
  - @pnpm/cli-utils@2.1.1

## 3.0.1

### Patch Changes

- @pnpm/license-scanner@2.2.1

## 3.0.0

### Major Changes

- d6592964f: `rootProjectManifestDir` is a required field.

### Minor Changes

- 43ce9e4a6: Support for multiple architectures when installing dependencies [#5965](https://github.com/pnpm/pnpm/issues/5965).

  You can now specify architectures for which you'd like to install optional dependencies, even if they don't match the architecture of the system running the install. Use the `supportedArchitectures` field in `package.json` to define your preferences.

  For example, the following configuration tells pnpm to install optional dependencies for Windows x64:

  ```json
  {
    "pnpm": {
      "supportedArchitectures": {
        "os": ["win32"],
        "cpu": ["x64"]
      }
    }
  }
  ```

  Whereas this configuration will have pnpm install optional dependencies for Windows, macOS, and the architecture of the system currently running the install. It includes artifacts for both x64 and arm64 CPUs:

  ```json
  {
    "pnpm": {
      "supportedArchitectures": {
        "os": ["win32", "darwin", "current"],
        "cpu": ["x64", "arm64"]
      }
    }
  }
  ```

  Additionally, `supportedArchitectures` also supports specifying the `libc` of the system.

### Patch Changes

- Updated dependencies [43ce9e4a6]
- Updated dependencies [d6592964f]
  - @pnpm/license-scanner@2.2.0
  - @pnpm/cli-utils@2.1.0
  - @pnpm/config@20.1.0
  - @pnpm/lockfile-file@8.1.4

## 2.1.0

### Minor Changes

- fff7866f3: The `pnpm licenses list` command now accepts the `--filter` option to check the licenses of the dependencies of a subset of workspace projects [#5806](https://github.com/pnpm/pnpm/issues/5806).

### Patch Changes

- Updated dependencies [fff7866f3]
  - @pnpm/license-scanner@2.1.0

## 2.0.30

### Patch Changes

- Updated dependencies [ac5abd3ff]
- Updated dependencies [b60bb6cbe]
  - @pnpm/config@20.0.0
  - @pnpm/license-scanner@2.0.22
  - @pnpm/cli-utils@2.0.24

## 2.0.29

### Patch Changes

- Updated dependencies [b1dd0ee58]
  - @pnpm/config@19.2.1
  - @pnpm/cli-utils@2.0.23

## 2.0.28

### Patch Changes

- Updated dependencies [d774a3196]
- Updated dependencies [832e28826]
  - @pnpm/config@19.2.0
  - @pnpm/cli-utils@2.0.22
  - @pnpm/lockfile-file@8.1.3
  - @pnpm/license-scanner@2.0.21

## 2.0.27

### Patch Changes

- Updated dependencies [ee328fd25]
  - @pnpm/config@19.1.0
  - @pnpm/cli-utils@2.0.21
  - @pnpm/license-scanner@2.0.20

## 2.0.26

### Patch Changes

- @pnpm/cli-utils@2.0.20

## 2.0.25

### Patch Changes

- @pnpm/config@19.0.3
- @pnpm/license-scanner@2.0.19
- @pnpm/cli-utils@2.0.19

## 2.0.24

### Patch Changes

- @pnpm/config@19.0.2
- @pnpm/license-scanner@2.0.18
- @pnpm/cli-utils@2.0.18

## 2.0.23

### Patch Changes

- @pnpm/license-scanner@2.0.17
- @pnpm/config@19.0.1

## 2.0.22

### Patch Changes

- @pnpm/license-scanner@2.0.16
- @pnpm/config@19.0.1
- @pnpm/cli-utils@2.0.17

## 2.0.21

### Patch Changes

- Updated dependencies [cb8bcc8df]
  - @pnpm/config@19.0.0
  - @pnpm/license-scanner@2.0.15
  - @pnpm/cli-utils@2.0.16

## 2.0.20

### Patch Changes

- @pnpm/cli-utils@2.0.15

## 2.0.19

### Patch Changes

- @pnpm/cli-utils@2.0.14

## 2.0.18

### Patch Changes

- @pnpm/license-scanner@2.0.14
- @pnpm/config@18.4.4

## 2.0.17

### Patch Changes

- @pnpm/license-scanner@2.0.13
- @pnpm/config@18.4.4

## 2.0.16

### Patch Changes

- @pnpm/license-scanner@2.0.12
- @pnpm/config@18.4.4

## 2.0.15

### Patch Changes

- @pnpm/cli-utils@2.0.13
- @pnpm/config@18.4.4
- @pnpm/lockfile-file@8.1.2
- @pnpm/license-scanner@2.0.11

## 2.0.14

### Patch Changes

- @pnpm/cli-utils@2.0.12
- @pnpm/config@18.4.3
- @pnpm/license-scanner@2.0.10

## 2.0.13

### Patch Changes

- @pnpm/license-scanner@2.0.9
- @pnpm/config@18.4.2

## 2.0.12

### Patch Changes

- c686768f0: `pnpm license ls` should work even when there is a patched git protocol dependency [#6595](https://github.com/pnpm/pnpm/issues/6595)
- Updated dependencies [e2d631217]
- Updated dependencies [c686768f0]
  - @pnpm/config@18.4.2
  - @pnpm/license-scanner@2.0.8
  - @pnpm/cli-utils@2.0.11

## 2.0.11

### Patch Changes

- Updated dependencies [302ebffc5]
  - @pnpm/constants@7.1.1
  - @pnpm/license-scanner@2.0.7
  - @pnpm/config@18.4.1
  - @pnpm/lockfile-file@8.1.1
  - @pnpm/error@5.0.2
  - @pnpm/cli-utils@2.0.10

## 2.0.10

### Patch Changes

- Updated dependencies [4b97f1f07]
  - @pnpm/license-scanner@2.0.6
  - @pnpm/config@18.4.0

## 2.0.9

### Patch Changes

- Updated dependencies [9c4ae87bd]
- Updated dependencies [9c4ae87bd]
- Updated dependencies [9c4ae87bd]
- Updated dependencies [301b8e2da]
  - @pnpm/lockfile-file@8.1.0
  - @pnpm/constants@7.1.0
  - @pnpm/config@18.4.0
  - @pnpm/license-scanner@2.0.5
  - @pnpm/cli-utils@2.0.9
  - @pnpm/error@5.0.1

## 2.0.8

### Patch Changes

- Updated dependencies [ee429b300]
- Updated dependencies [1de07a4af]
  - @pnpm/cli-utils@2.0.8
  - @pnpm/config@18.3.2
  - @pnpm/license-scanner@2.0.4

## 2.0.7

### Patch Changes

- Updated dependencies [2809e89ab]
  - @pnpm/config@18.3.1
  - @pnpm/cli-utils@2.0.7

## 2.0.6

### Patch Changes

- Updated dependencies [32f8e08c6]
- Updated dependencies [c0760128d]
  - @pnpm/config@18.3.0
  - @pnpm/lockfile-file@8.0.2
  - @pnpm/cli-utils@2.0.6
  - @pnpm/license-scanner@2.0.3

## 2.0.5

### Patch Changes

- Updated dependencies [fc8780ca9]
  - @pnpm/config@18.2.0
  - @pnpm/license-scanner@2.0.2
  - @pnpm/cli-utils@2.0.5

## 2.0.4

### Patch Changes

- Updated dependencies [5087636b6]
- Updated dependencies [94f94eed6]
  - @pnpm/lockfile-file@8.0.1
  - @pnpm/license-scanner@2.0.1
  - @pnpm/cli-utils@2.0.4
  - @pnpm/config@18.1.1

## 2.0.3

### Patch Changes

- Updated dependencies [e2cb4b63d]
- Updated dependencies [cd6ce11f0]
  - @pnpm/config@18.1.0
  - @pnpm/cli-utils@2.0.3

## 2.0.2

### Patch Changes

- @pnpm/config@18.0.2
- @pnpm/cli-utils@2.0.2

## 2.0.1

### Patch Changes

- @pnpm/config@18.0.1
- @pnpm/cli-utils@2.0.1

## 2.0.0

### Major Changes

- eceaa8b8b: Node.js 14 support dropped.

### Patch Changes

- 22ccf155e: Fix `Segmentation fault` error in the bundled version of pnpm [#6241](https://github.com/pnpm/pnpm/issues/6241).
- Updated dependencies [47e45d717]
- Updated dependencies [47e45d717]
- Updated dependencies [158d8cf22]
- Updated dependencies [eceaa8b8b]
- Updated dependencies [8e35c21d1]
- Updated dependencies [47e45d717]
- Updated dependencies [47e45d717]
- Updated dependencies [113f0ae26]
- Updated dependencies [417c8ac59]
  - @pnpm/config@18.0.0
  - @pnpm/lockfile-file@8.0.0
  - @pnpm/common-cli-options-help@1.0.0
  - @pnpm/license-scanner@2.0.0
  - @pnpm/constants@7.0.0
  - @pnpm/store-path@8.0.0
  - @pnpm/error@5.0.0
  - @pnpm/cli-utils@2.0.0
  - @pnpm/command@4.0.0

## 1.0.33

### Patch Changes

- @pnpm/config@17.0.2
- @pnpm/cli-utils@1.1.7

## 1.0.32

### Patch Changes

- Updated dependencies [b38d711f3]
  - @pnpm/config@17.0.1
  - @pnpm/cli-utils@1.1.6

## 1.0.31

### Patch Changes

- Updated dependencies [787c43dcc]
- Updated dependencies [e505b58e3]
  - @pnpm/lockfile-file@7.0.6
  - @pnpm/config@17.0.0
  - @pnpm/license-scanner@1.0.17
  - @pnpm/cli-utils@1.1.5

## 1.0.30

### Patch Changes

- @pnpm/config@16.7.2
- @pnpm/cli-utils@1.1.4

## 1.0.29

### Patch Changes

- 019e4f2de: Should not throw an error when local dependency use file protocol [#6115](https://github.com/pnpm/pnpm/issues/6115).
- Updated dependencies [019e4f2de]
  - @pnpm/license-scanner@1.0.16
  - @pnpm/config@16.7.1
  - @pnpm/cli-utils@1.1.3

## 1.0.28

### Patch Changes

- Updated dependencies [7d64d757b]
- Updated dependencies [5c31fa8be]
  - @pnpm/cli-utils@1.1.2
  - @pnpm/config@16.7.0

## 1.0.27

### Patch Changes

- Updated dependencies [ed946c73e]
  - @pnpm/lockfile-file@7.0.5
  - @pnpm/license-scanner@1.0.15
  - @pnpm/config@16.6.4
  - @pnpm/cli-utils@1.1.1

## 1.0.26

### Patch Changes

- Updated dependencies [0377d9367]
  - @pnpm/cli-utils@1.1.0
  - @pnpm/config@16.6.3

## 1.0.25

### Patch Changes

- @pnpm/config@16.6.2
- @pnpm/cli-utils@1.0.34

## 1.0.24

### Patch Changes

- @pnpm/config@16.6.1
- @pnpm/license-scanner@1.0.14
- @pnpm/cli-utils@1.0.33

## 1.0.23

### Patch Changes

- Updated dependencies [59ee53678]
  - @pnpm/config@16.6.0
  - @pnpm/lockfile-file@7.0.4
  - @pnpm/license-scanner@1.0.13
  - @pnpm/cli-utils@1.0.32

## 1.0.22

### Patch Changes

- @pnpm/lockfile-file@7.0.3
- @pnpm/license-scanner@1.0.12
- @pnpm/config@16.5.5
- @pnpm/cli-utils@1.0.31

## 1.0.21

### Patch Changes

- @pnpm/config@16.5.4
- @pnpm/cli-utils@1.0.30

## 1.0.20

### Patch Changes

- @pnpm/config@16.5.3
- @pnpm/cli-utils@1.0.29

## 1.0.19

### Patch Changes

- @pnpm/config@16.5.2
- @pnpm/cli-utils@1.0.28

## 1.0.18

### Patch Changes

- @pnpm/license-scanner@1.0.11
- @pnpm/config@16.5.1
- @pnpm/cli-utils@1.0.27

## 1.0.17

### Patch Changes

- Updated dependencies [28b47a156]
  - @pnpm/config@16.5.0
  - @pnpm/cli-utils@1.0.26

## 1.0.16

### Patch Changes

- @pnpm/license-scanner@1.0.10
- @pnpm/config@16.4.3
- @pnpm/cli-utils@1.0.25

## 1.0.15

### Patch Changes

- Updated dependencies [9a68ebbae]
  - @pnpm/lockfile-file@7.0.2
  - @pnpm/license-scanner@1.0.9
  - @pnpm/config@16.4.2
  - @pnpm/cli-utils@1.0.24

## 1.0.14

### Patch Changes

- @pnpm/lockfile-file@7.0.1
- @pnpm/license-scanner@1.0.8
- @pnpm/config@16.4.1
- @pnpm/cli-utils@1.0.23

## 1.0.13

### Patch Changes

- Updated dependencies [3ebce5db7]
- Updated dependencies [3ebce5db7]
- Updated dependencies [3ebce5db7]
  - @pnpm/constants@6.2.0
  - @pnpm/lockfile-file@7.0.0
  - @pnpm/config@16.4.0
  - @pnpm/error@4.0.1
  - @pnpm/license-scanner@1.0.7
  - @pnpm/cli-utils@1.0.22

## 1.0.12

### Patch Changes

- Updated dependencies [1fad508b0]
  - @pnpm/config@16.3.0
  - @pnpm/cli-utils@1.0.21

## 1.0.11

### Patch Changes

- @pnpm/cli-utils@1.0.20
- @pnpm/config@16.2.2

## 1.0.10

### Patch Changes

- Updated dependencies [d71dbf230]
  - @pnpm/config@16.2.1
  - @pnpm/cli-utils@1.0.19

## 1.0.9

### Patch Changes

- 1d3995fe3: Add the 'description'-field to the licenses output [#5836](https://github.com/pnpm/pnpm/pull/5836).
- Updated dependencies [1d3995fe3]
- Updated dependencies [841f52e70]
  - @pnpm/license-scanner@1.0.6
  - @pnpm/config@16.2.0
  - @pnpm/cli-utils@1.0.18

## 1.0.8

### Patch Changes

- @pnpm/cli-utils@1.0.17
- @pnpm/config@16.1.11
- @pnpm/lockfile-file@6.0.5
- @pnpm/license-scanner@1.0.5

## 1.0.7

### Patch Changes

- @pnpm/lockfile-file@6.0.4
- @pnpm/license-scanner@1.0.4
- @pnpm/config@16.1.10
- @pnpm/cli-utils@1.0.16

## 1.0.6

### Patch Changes

- 5464e1da6: `pnpm license list` should not fail if a license file is an executable [#5740](https://github.com/pnpm/pnpm/pull/5740).
- Updated dependencies [5464e1da6]
  - @pnpm/license-scanner@1.0.3
  - @pnpm/config@16.1.9
  - @pnpm/cli-utils@1.0.15

## 1.0.5

### Patch Changes

- 568dc3ab2: `pnpm licenses` should print help, not just an error message [#5745](https://github.com/pnpm/pnpm/issues/5745).
  - @pnpm/cli-utils@1.0.14
  - @pnpm/config@16.1.8

## 1.0.4

### Patch Changes

- Updated dependencies [a9d59d8bc]
  - @pnpm/config@16.1.7
  - @pnpm/lockfile-file@6.0.3
  - @pnpm/cli-utils@1.0.13
  - @pnpm/license-scanner@1.0.2

## 1.0.3

### Patch Changes

- @pnpm/config@16.1.6
- @pnpm/cli-utils@1.0.12

## 1.0.2

### Patch Changes

- @pnpm/config@16.1.5
- @pnpm/cli-utils@1.0.11

## 1.0.1

### Patch Changes

- a8cc22364: Fix the CLI help of the `pnpm licenses` command.
  - @pnpm/cli-utils@1.0.10
  - @pnpm/config@16.1.4
  - @pnpm/license-scanner@1.0.1

## 1.0.0

### Major Changes

- d84a30a04: Added a new command `pnpm licenses list`, which displays the licenses of the packages [#2825](https://github.com/pnpm/pnpm/issues/2825)

### Patch Changes

- Updated dependencies [d84a30a04]
  - @pnpm/license-scanner@1.0.0
  - @pnpm/config@16.1.3
  - @pnpm/cli-utils@1.0.9
