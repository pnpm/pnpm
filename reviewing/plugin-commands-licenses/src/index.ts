import * as licenses from './licenses.js'

export { licenses }
