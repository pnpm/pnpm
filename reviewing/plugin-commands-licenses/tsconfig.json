{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../__typings__/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../../__utils__/prepare"
    },
    {
      "path": "../../__utils__/test-fixtures"
    },
    {
      "path": "../../cli/cli-utils"
    },
    {
      "path": "../../cli/command"
    },
    {
      "path": "../../cli/common-cli-options-help"
    },
    {
      "path": "../../config/config"
    },
    {
      "path": "../../lockfile/lockfile-file"
    },
    {
      "path": "../../packages/constants"
    },
    {
      "path": "../../packages/error"
    },
    {
      "path": "../../pkg-manager/plugin-commands-installation"
    },
    {
      "path": "../../pkg-manifest/read-package-json"
    },
    {
      "path": "../../store/store-path"
    },
    {
      "path": "../../workspace/filter-workspace-packages"
    },
    {
      "path": "../license-scanner"
    }
  ],
  "composite": true
}
