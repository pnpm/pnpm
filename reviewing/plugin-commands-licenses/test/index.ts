/// <reference path="../../../__typings__/index.d.ts" />
import path from 'path'
import { licenses } from '@pnpm/plugin-commands-licenses'
import { install } from '@pnpm/plugin-commands-installation'
import { tempDir } from '@pnpm/prepare'
import { fixtures } from '@pnpm/test-fixtures'
import stripAnsi from 'strip-ansi'
import { DEFAULT_OPTS } from './utils'
import { readProjects } from '@pnpm/filter-workspace-packages'

const f = fixtures(__dirname)

test('pnpm licenses', async () => {
  const workspaceDir = tempDir()
  f.copy('complex-licenses', workspaceDir)

  const storeDir = path.join(workspaceDir, 'store')
  await install.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    storeDir,
  })

  // Attempt to run the licenses command now
  const { output, exitCode } = await licenses.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    long: false,
    // we need to prefix it with v3 otherwise licenses tool can't find anything
    // in the content-addressable directory
    storeDir: path.resolve(storeDir, 'v3'),
  }, ['list'])

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toMatchSnapshot('show-packages')
})

test('pnpm licenses: show details', async () => {
  const workspaceDir = tempDir()
  f.copy('simple-licenses', workspaceDir)

  const storeDir = path.join(workspaceDir, 'store')
  await install.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    storeDir,
  })

  // Attempt to run the licenses command now
  const { output, exitCode } = await licenses.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    long: true,
    // we need to prefix it with v3 otherwise licenses tool can't find anything
    // in the content-addressable directory
    storeDir: path.resolve(storeDir, 'v3'),
  }, ['list'])

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toMatchSnapshot('show-packages-details')
})

test('pnpm licenses: output as json', async () => {
  const workspaceDir = tempDir()
  f.copy('simple-licenses', workspaceDir)

  const storeDir = path.join(workspaceDir, 'store')
  await install.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    storeDir,
  })

  // Attempt to run the licenses command now
  const { output, exitCode } = await licenses.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    long: false,
    json: true,
    // we need to prefix it with v3 otherwise licenses tool can't find anything
    // in the content-addressable directory
    storeDir: path.resolve(storeDir, 'v3'),
  }, ['list'])

  expect(exitCode).toBe(0)
  expect(output).not.toHaveLength(0)
  expect(output).not.toBe('No licenses in packages found')
  const parsedOutput = JSON.parse(output)
  expect(Object.keys(parsedOutput)).toMatchSnapshot('found-license-types')
  const packagesWithMIT = parsedOutput['MIT']
  expect(packagesWithMIT.length).toBeGreaterThan(0)
  expect(Object.keys(packagesWithMIT[0])).toEqual([
    'name',
    'version',
    'path',
    'license',
    'author',
    'homepage',
    'description',
  ])
  expect(packagesWithMIT[0].name).toBe('is-positive')
})

test('pnpm licenses: filter outputs', async () => {
  const workspaceDir = tempDir()
  f.copy('workspace-licenses', workspaceDir)

  const { allProjects, allProjectsGraph, selectedProjectsGraph } =
    await readProjects(workspaceDir, [])

  const storeDir = path.join(workspaceDir, 'store')
  await install.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    workspaceDir,
    lockfileDir: workspaceDir,
    pnpmHomeDir: '',
    storeDir,
    allProjects,
    allProjectsGraph,
    selectedProjectsGraph,
  })

  const { output, exitCode } = await licenses.handler(
    {
      ...DEFAULT_OPTS,
      dir: workspaceDir,
      pnpmHomeDir: '',
      long: false,
      selectedProjectsGraph: Object.fromEntries(
        Object.entries(selectedProjectsGraph).filter(([path]) =>
          path.includes('bar')
        )
      ),
      storeDir: path.resolve(storeDir, 'v3'),
    }, ['list']
  )

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toMatchSnapshot('show-packages')
})

test('pnpm licenses: fails when lockfile is missing', async () => {
  await expect(
    licenses.handler({
      ...DEFAULT_OPTS,
      dir: path.resolve('./test/fixtures/invalid'),
      pnpmHomeDir: '',
      long: true,
    }, ['list'])
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    '"No pnpm-lock.yaml found: Cannot check a project without a lockfile"'
  )
})

test('pnpm licenses: should correctly read LICENSE file with executable file mode', async () => {
  const workspaceDir = tempDir()
  f.copy('file-mode-test', workspaceDir)

  const storeDir = path.join(workspaceDir, 'store')
  await install.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    storeDir,
  })

  // Attempt to run the licenses command now
  const { output, exitCode } = await licenses.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    long: true,
    // we need to prefix it with v3 otherwise licenses tool can't find anything
    // in the content-addressable directory
    storeDir: path.resolve(storeDir, 'v3'),
  }, ['list'])

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toMatchSnapshot('show-packages-details')
})

test('pnpm licenses should work with file protocol dependency', async () => {
  const workspaceDir = tempDir()
  f.copy('with-file-protocol', workspaceDir)

  const storeDir = path.join(workspaceDir, 'store')
  await install.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    storeDir,
  })

  const { output, exitCode } = await licenses.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    long: false,
    storeDir: path.resolve(storeDir, 'v3'),
  }, ['list'])

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toMatchSnapshot('show-packages')
})

test('pnpm licenses should work with git protocol dep that have patches', async () => {
  const workspaceDir = tempDir()
  f.copy('with-git-protocol-patched-deps', workspaceDir)

  const storeDir = path.join(workspaceDir, 'store')
  await install.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    storeDir,
  })

  const { exitCode } = await licenses.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    long: false,
    storeDir: path.resolve(storeDir, 'v3'),
  }, ['list'])

  expect(exitCode).toBe(0)
})

test('pnpm licenses should work with git protocol dep that have peerDependencies', async () => {
  const workspaceDir = tempDir()
  f.copy('with-git-protocol-peer-deps', workspaceDir)

  const storeDir = path.join(workspaceDir, 'store')
  await install.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    storeDir,
  })

  const { exitCode } = await licenses.handler({
    ...DEFAULT_OPTS,
    dir: workspaceDir,
    pnpmHomeDir: '',
    long: false,
    storeDir: path.resolve(storeDir, 'v3'),
  }, ['list'])

  expect(exitCode).toBe(0)
})
