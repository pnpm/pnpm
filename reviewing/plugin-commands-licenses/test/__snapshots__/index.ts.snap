// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pnpm licenses should work with file protocol dependency: show-packages 1`] = `
"┌─────────────────────────────────┬─────────┐
│ Package                         │ License │
├─────────────────────────────────┼─────────┤
│ is-positive                     │ MIT     │
├─────────────────────────────────┼─────────┤
│ pnpm-with-file-protocol-sub-dep │ Unknown │
└─────────────────────────────────┴─────────┘
"
`;

exports[`pnpm licenses: filter outputs: show-packages 1`] = `
"┌─────────────┬─────────┐
│ Package     │ License │
├─────────────┼─────────┤
│ is-positive │ MIT     │
└─────────────┴─────────┘
"
`;

exports[`pnpm licenses: should correctly read LICENSE file with executable file mode: show-packages-details 1`] = `
"┌──────────────────────────────┬─────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Package                      │ License │ Details                                                                                                │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ inherits                     │ ISC     │ Browser-friendly inheritance fully compatible with standard node.js inherits()                         │
│                              │         │ https://github.com/isaacs/inherits#readme                                                              │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ sax                          │ ISC     │ Isaac Z. Schlueter                                                                                     │
│                              │         │ An evented streaming XML parser in JavaScript                                                          │
│                              │         │ https://github.com/isaacs/sax-js#readme                                                                │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ commander                    │ MIT     │ TJ Holowaychuk                                                                                         │
│                              │         │ the complete solution for node.js command-line programs                                                │
│                              │         │ https://github.com/tj/commander.js#readme                                                              │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ core-util-is                 │ MIT     │ Isaac Z. Schlueter                                                                                     │
│                              │         │ The \`util.is*\` functions introduced in Node v0.12.                                                     │
│                              │         │ https://github.com/isaacs/core-util-is#readme                                                          │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ isarray                      │ MIT     │ Julian Gruber                                                                                          │
│                              │         │ Array#isArray for older browsers                                                                       │
│                              │         │ https://github.com/juliangruber/isarray                                                                │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ neatequal                    │ MIT     │ Nicolas Froidure                                                                                       │
│                              │         │ Neat deep equal.                                                                                       │
│                              │         │ https://github.com/nfroidure/neatequal                                                                 │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ process-nextick-args         │ MIT     │ process.nextTick but always with args                                                                  │
│                              │         │ https://github.com/calvinmetcalf/process-nextick-args                                                  │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ readable-stream              │ MIT     │ Streams3, a user-land copy of the stream library from Node.js                                          │
│                              │         │ https://github.com/nodejs/readable-stream#readme                                                       │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ safe-buffer                  │ MIT     │ Feross Aboukhadijeh                                                                                    │
│                              │         │ Safer Node.js Buffer API                                                                               │
│                              │         │ https://github.com/feross/safe-buffer                                                                  │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ string_decoder               │ MIT     │ The string_decoder module from Node core                                                               │
│                              │         │ https://github.com/nodejs/string_decoder                                                               │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ string.fromcodepoint         │ MIT     │ Mathias Bynens                                                                                         │
│                              │         │ A robust & optimized \`String.fromCodePoint\` polyfill, based on the ECMAScript 6 specification.         │
│                              │         │ http://mths.be/fromcodepoint                                                                           │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ string.prototype.codepointat │ MIT     │ Mathias Bynens                                                                                         │
│                              │         │ A robust & optimized \`String.prototype.codePointAt\` polyfill, based on the ECMAScript 6 specification. │
│                              │         │ https://mths.be/codepointat                                                                            │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ svg-pathdata                 │ MIT     │ Nicolas Froidure                                                                                       │
│                              │         │ Parse, transform and encode SVG Path Data.                                                             │
│                              │         │ https://github.com/nfroidure/SVGPathData#readme                                                        │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ svgicons2svgfont             │ MIT     │ Nicolas Froidure                                                                                       │
│                              │         │ Read a set of SVG icons and ouput a SVG font                                                           │
│                              │         │ https://github.com/nfroidure/svgicons2svgfont                                                          │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ util-deprecate               │ MIT     │ Nathan Rajlich                                                                                         │
│                              │         │ The Node.js \`util.deprecate()\` function with browser support                                           │
│                              │         │ https://github.com/TooTallNate/util-deprecate                                                          │
├──────────────────────────────┼─────────┼────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ varstream                    │ MIT     │ Nicolas Froidure                                                                                       │
│                              │         │ Stream variables beetween 2 JavaScript threads (client/server, ipc, worker/main thread).               │
│                              │         │ https://github.com/nfroidure/VarStream#readme                                                          │
└──────────────────────────────┴─────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────┘
"
`;

exports[`pnpm licenses: show details: show-packages-details 1`] = `
"┌─────────────┬─────────┬─────────────────────────────────────────────┐
│ Package     │ License │ Details                                     │
├─────────────┼─────────┼─────────────────────────────────────────────┤
│ is-positive │ MIT     │ Kevin Martensson                            │
│             │         │ Check if something is a positive number     │
│             │         │ https://github.com/kevva/is-positive#readme │
└─────────────┴─────────┴─────────────────────────────────────────────┘
"
`;

exports[`pnpm licenses: show-packages 1`] = `
"┌──────────────────┬────────────┐
│ Package          │ License    │
├──────────────────┼────────────┤
│ typescript (dev) │ Apache-2.0 │
├──────────────────┼────────────┤
│ js-tokens        │ MIT        │
├──────────────────┼────────────┤
│ loose-envify     │ MIT        │
├──────────────────┼────────────┤
│ react            │ MIT        │
├──────────────────┼────────────┤
│ react-dom        │ MIT        │
├──────────────────┼────────────┤
│ scheduler        │ MIT        │
└──────────────────┴────────────┘
"
`;
